### PR TITLE
SEV-SNP ACI: Store and cross-verify UVM endorsements

### PR DESCRIPTION
**Pull request by jumaffre**
Resolves https://github.com/microsoft/CCF/issues/4193

Follow-up from https://github.com/microsoft/CCF/pull/4915. Now that we verify UVM endorsements on node startup, this PR adds support for record of the associated `{did, feed, svn}` in the ledger, as well as verification of new joining nodes which now also provided their own UVM endorsements that are verified as part of the join protocol.

### CODE REVIEW

File: doc/audit/builtin_maps.rst [hunk 1]
```diff
@@ -166,6 +168,15 @@ Measurement that nodes can run and join the current network on :doc:`SNP <../ope
    * - ``ede8268...01b66ed1``
      - ``ALLOWED_TO_JOIN``
 
+``nodes.snp.uvm_endorsements``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+For Confidential Azure Container Instance (ACI) deployments, trusted endorsements of utility VM (UVM) for new nodes allowed to join the network (:doc:`SNP <../operations/platforms/snp>` only).
+
+**Key** Trusted endorser DID.
```
Comment (achamayou) [inline doc/audit/builtin_maps.rst:176]:
```suggestion
**Key** Trusted endorser DID (did:x509 only for now: https://github.com/microsoft/did-x509/blob/main/specification.md).
```

File: doc/audit/builtin_maps.rst [hunk 2]
```diff
@@ -166,6 +168,15 @@ Measurement that nodes can run and join the current network on :doc:`SNP <../ope
...
+
+**Value** Map of issuer feed to Security Version Number (SVN) represented as JSON.
```
Comment (achamayou) [inline doc/audit/builtin_maps.rst:178]:
```suggestion
**Value** Map of issuer feed to Security Version Number (SVN) represented as JSON. See https://ietf-wg-scitt.github.io/draft-ietf-scitt-architecture/draft-ietf-scitt-architecture.html#name-issuer-identity.
```

File: samples/constitutions/default/actions.js [hunk 3]
```diff
@@ -977,6 +977,32 @@ const actions = new Map([
       }
     ),
   ],
+  [
+    "add_snp_uvm_endorsement",
+    new Action(
+      function (args) {
+        checkType(args.did, "string", "did");
+        checkType(args.feed, "string", "feed");
+        checkType(args.svn, "integer", "svn");
```
Comment (achamayou) [inline samples/constitutions/default/actions.js:986]: We should check that the value is positive too.

File: src/node/quote.cpp [hunk 4]
```diff
@@ -6,32 +6,84 @@
...
 namespace ccf
 {
+  bool verify_enclave_measurement_against_uvm_endorsements(
+    kv::ReadOnlyTx& tx,
+    const CodeDigest& quote_measurement,
+    const std::vector<uint8_t>& uvm_endorsements)
+  {
+    auto uvm_endorsements_data =
+      verify_uvm_endorsements(uvm_endorsements, quote_measurement);
+    auto uvmes = tx.ro<SnpUVMEndorsements>(Tables::NODE_SNP_UVM_ENDORSEMENTS);
```
Comment (achamayou) [inline src/node/quote.cpp:21]: It's weird to have SnpUVM, I think we can either have SnpUvm, or SNPUVM, but the mix seems quite inconsistent.

File: src/node/uvm_endorsements.h [hunk 5]
```diff
@@ -301,8 +305,12 @@ namespace ccf
 
     LOG_INFO_FMT(
       "Successfully verified endorsements for attested measurement {} against "
-      "{}",
+      "{}, feed {}, svn {}",
       payload.sevsnpvm_launch_measurement,
-      did);
+      did,
+      phdr.feed,
+      payload.sevsnpvn_guest_svn);
```
Comment (achamayou) [inline src/node/uvm_endorsements.h:312]:
```suggestion
      payload.sevsnp_guest_svn);
```
I think?

**Review by achamayou — APPROVED** [review]

**Review by DomAyre — APPROVED** [review]

**Pull request merged**